CachedBlobContainerIndexInputTests.testRandomReads failure

**Build scan**: https://gradle-enterprise.elastic.co/s/tiqw5pjr4tiwk

**Repro line**:
```
./gradlew ':x-pack:plugin:searchable-snapshots:test' --tests "org.elasticsearch.index.store.cache.CachedBlobContainerIndexInputTests.testRandomReads" -Dtests.seed=9EFCEB28234D764F -Dtests.security.manager=true -Dtests.locale=en-AU -Dtests.timezone=Pacific/Auckland -Druntime.java=8
```

**Reproduces locally?**:
No.

**Applicable branches**:

At least 7.x, that is where it failed on #64160 

**Failure history**:

Looks like it only failed [once](https://build-stats.elastic.co/app/kibana#/discover?_g=(refreshInterval:(pause:!t,value:0),time:(from:now-30d,mode:quick,to:now))&_a=(columns:!(_source),index:b646ed00-7efc-11e8-bf69-63c8ef516157,interval:auto,query:(language:lucene,query:testRandomReads),sort:!(process.time-start,desc))), though there is also a failure of the test from September 3rd (looks like an unrelated failure).

**Failure excerpt**:

```
java.lang.AssertionError: All bytes should have been read from source	
Expected: <20522L>	
     but: was <20115L>
     at __randomizedtesting.SeedInfo.seed([9EFCEB28234D764F:ABFE49DBC9E1D265]:0)
     at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
     at org.junit.Assert.assertThat(Assert.java:956)
     at org.elasticsearch.index.store.cache.CachedBlobContainerIndexInputTests.testRandomReads(CachedBlobContainerIndexInputTests.java:150)
```

--- a/x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/index/store/cache/CachedBlobContainerIndexInputTests.java
+++ b/x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/index/store/cache/CachedBlobContainerIndexInputTests.java
@@ -45,6 +45,7 @@ import java.nio.charset.StandardCharsets;
 import java.nio.file.Path;
 import java.util.HashSet;
 import java.util.Objects;
+import java.util.concurrent.atomic.AtomicInteger;
 import java.util.concurrent.atomic.LongAdder;
 
 import static java.util.Collections.singletonList;
@@ -151,6 +152,14 @@ public class CachedBlobContainerIndexInputTests extends ESIndexInputTestCase {
                         "All bytes should have been read from source",
                         ((CountingBlobContainer) blobContainer).totalBytes.sum(),
                         equalTo((long) input.length)
+                    );
+                    // busy assert that closing of all streams happened because they are closed on background fetcher threads
+                    assertBusy(
+                        () -> assertEquals(
+                            "All open streams should have been closed",
+                            0,
+                            ((CountingBlobContainer) blobContainer).openStreams.get()
+                        )
                     );
                 }
             }
@@ -269,6 +278,8 @@ public class CachedBlobContainerIndexInputTests extends ESIndexInputTestCase {
         private final LongAdder totalBytes = new LongAdder();
         private final LongAdder totalOpens = new LongAdder();
 
+        private final AtomicInteger openStreams = new AtomicInteger(0);
+
         private final int rangeSize;
 
         CountingBlobContainer(BlobContainer in, int rangeSize) {
@@ -301,7 +312,6 @@ public class CachedBlobContainerIndexInputTests extends ESIndexInputTestCase {
 
         private final CountingBlobContainer container;
 
-        private long bytesRead = 0L;
         private long position = 0L;
         private long start = Long.MAX_VALUE;
         private long end = Long.MIN_VALUE;
@@ -310,6 +320,7 @@ public class CachedBlobContainerIndexInputTests extends ESIndexInputTestCase {
             super(input);
             this.container = Objects.requireNonNull(container);
             this.container.totalOpens.increment();
+            this.container.openStreams.incrementAndGet();
         }
 
         @Override
@@ -322,7 +333,7 @@ public class CachedBlobContainerIndexInputTests extends ESIndexInputTestCase {
             if (result == -1) {
                 return result;
             }
-            bytesRead += 1L;
+            this.container.totalBytes.increment();
             position += 1L;
 
             if (position > end) {
@@ -338,7 +349,7 @@ public class CachedBlobContainerIndexInputTests extends ESIndexInputTestCase {
             }
 
             final int result = in.read(b, offset, len);
-            bytesRead += len;
+            this.container.totalBytes.add(len);
             position += len;
 
             if (position > end) {
@@ -349,8 +360,8 @@ public class CachedBlobContainerIndexInputTests extends ESIndexInputTestCase {
 
         @Override
         public void close() throws IOException {
-            in.close();
-            this.container.totalBytes.add(bytesRead);
+            super.close();
+            this.container.openStreams.decrementAndGet();
         }
     }
 }

